### ISSUE TITLE
Pageview is sent twice on the same page

### ISSUE THREAD

**Issue by onklmaps** [PrestaShop/ps_googleanalytics#41]
`ga('send', 'pageview’);` is echoed twice on front controller.
Once in `ps_googleanalytics/views/templates/hook/ps_googleanalytics.tpl:57` and another time in:
`ps_googleanalytics/ps_googleanalytics.php:747`.

ps_googleanalytics 3.2.0
ps 1.7.5.2

**Comment by khouloudbelguith** [comment]
Related to https://github.com/PrestaShop/ps_googleanalytics/pull/37

**Comment by jaxtheking** [comment]
Same for me on latest 1.7.6.3. I realised there was a problem when I saw my bounce rate going down to almost 0 in Google Analytics.

**Comment by sandor111** [comment]
![Screenshot_111](https://user-images.githubusercontent.com/17690323/74769565-ef2eaf00-528a-11ea-8738-e87ab8ba73fe.png)
Same here

**Comment by Slews** [comment]
The pageview send order is indeed called twice:

- First in views/templates/hook/ps_googleanalytics.tpl line 57 if !$backoffice
- Then with the hookdisplayFooter() call. The hook execution starts with js_state to 0 (I don't understand the role of this property and so I don't know why / when it should be set to 0 or 1). This causes the next _runJs() call to add the second pageview call (line 749 of ps_googleanalytics.php).

For now, I am using a ps_googleanalytics.tpl in my theme so I can remove the first call without touching the module itself.

**Comment by webeshop** [comment]
Same problem is on 1.7.5.1 PrestaShop.

**Comment by sandor111** [comment]
Still happening

**Comment by dali-rajab** [comment]
you can make a temporary override until this issu is fixed. Make a file **override/modules/ps_googleanalytics/ps_googleanalytics.php**

```php
<?php
/**
 * 2007-2018 PrestaShop
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Academic Free License (AFL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/afl-3.0.php
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@prestashop.com so we can send you a copy immediately.
 *
 * DISCLAIMER
 *
 * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
 * versions in the future. If you wish to customize PrestaShop for your
 * needs please refer to http://www.prestashop.com for more information.
 *
 *  @author    PrestaShop SA <contact@prestashop.com>
 *  @copyright 2007-2018 PrestaShop SA
 *  @license   http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
 *  International Registered Trademark & Property of PrestaShop SA
 */

if (!defined('_PS_VERSION_')) {
    exit;
}

class Ps_GoogleanalyticsOverride extends Ps_Googleanalytics
{
    protected function _runJs($js_code, $backoffice = 0) {

        $original_runJs = parent::_runJs($js_code, $backoffice = 0);

        $pageviewPattern = '/ga\(\s*[\'"]send[\'"]\s*,\s*[\'"]pageview[\'"]\s*\)\s*;/im'; // this is the regex of ga('send', 'pageview');
        
        $filtered_runJs = preg_replace($pageviewPattern, '', $original_runJs); // remove ga('send', 'pageview'); from the original js

        return $filtered_runJs;
    }
    
    public function hookdisplayHeader($params, $back_office = false) {
        
        // The order of these lines is important!
        
        $original_return = parent::hookdisplayHeader($params, $back_office = false); // this will load GoogleAnalyticActionLib.js
        
        if (Configuration::get('GA_ACCOUNT_ID')) {
            $this->context->controller->addJs($this->_path.'views/js/GoogleAnalyticActionLibRectifier.js'); // this wil remove ga('send', 'pageview'); from GoogleAnalyticActionLib.js
        }
        
        return $original_return;
    }
}
```
And make another file in your active theme **themes/your_theme/modules/ps_googleanalytics/views/js/GoogleAnalyticActionLibRectifier.js**
```javascript
if (typeof GoogleAnalyticEnhancedECommerce !== 'undefined') {
  if (GoogleAnalyticEnhancedECommerce.hasOwnProperty('addCheckout')) {
    GoogleAnalyticEnhancedECommerce.addCheckout = function(Step) {
  		ga('ec:setAction', 'checkout', {
  			'step': Step
  		});
  	}
  }
}
```

**Comment by vitakrejci** [comment]
> you can make a temporary override until this issu is fixed. Make a file **override/modules/ps_googleanalytics/ps_googleanalytics.php**
> 
Thanks, it works!

**Comment by matks** [comment]
It seems issue is fixed in v4.0.0, I just checked it. Feel free to reopen an issue if you find usecase where issue is still observable.
